### PR TITLE
CI (Buildkite): Add the `package_linux64` and `doctest` builders

### PR DESCRIPTION
**Pull request by DilumAluthge**
TODO:
- [x] Don't start the `tester_linux64` builder until the `package_linux64` builder has completed successfully.
- [x] At the end of the `package_linux64` builder, upload the build artifacts.
- [x] At the beginning of the `tester_linux64` builder, download the build artifacts from `package_linux64` (instead of building Julia from source).
- [x] Get all doctests to pass in the `doctest` builder.

### CODE REVIEW

**Comment by DilumAluthge** [comment]
@staticfloat Take a look at the test failures. It looks like most (if not all) are due to issues with the sandboxing?

**Comment by staticfloat** [comment]
Wow, this is a lot of work Dilum!  Awesome to have you working on these pieces.  A few things to note:

1) We are going to want to coordinate GCC versions across platforms, so that we can, in general, say something like "Julia is built with GCC 9" and have that be true across all platforms.  To that end, we are going to want to more tightly control how the `package_linux64` image is built.  I don't think we should re-use the `llvm-passes` image, but should instead construct our own, and probably add a PPA/repository that allows us to add a specific version of GCC directly.

2) I've never tried running `rr` inside of `sandbox`, inside of `sandbox`; if that "just works" I will be ecstatic.

3) [`path.jl` test failure](https://buildkite.com/julialang/julia-master/builds/1861#2008d3d7-db08-4db1-93db-a8227f702b37/187-373) is due to `$HOME` not existing within the sandbox; we just need to set an appropriate environment mapping here.

4) [`file.jl` test failure](https://buildkite.com/julialang/julia-master/builds/1861#2008d3d7-db08-4db1-93db-a8227f702b37/187-412) is due to Julia mistakenly thinking we're `root` during the build process (because [either `$USER` or `$HOME` looks like it](https://github.com/JuliaLang/julia/blob/a9232d67203e6f1aca4934318d86f85cad50550b/test/file.jl#L496)) and trying to do a `chown()` that gets denied unless you're root.  We should not run as root, so we should change `$USER` and `$HOME` to be something less root-like.

5) [`file.jl` warning](https://buildkite.com/julialang/julia-master/builds/1861#2008d3d7-db08-4db1-93db-a8227f702b37/187-458) is also due to the `root` detection going awry.

A quick glance over the rest of the errors all look like they stem from `$HOME` not existing.  So let's fix these issues (by defining `HOME` in the yaml to point to some directory that exists within the sandbox, for one) and see if that improves things significantly.

**Comment by DilumAluthge** [comment]
> 2\. I've never tried running `rr` inside of `sandbox`, inside of `sandbox`; if that "just works" I will be ecstatic.

Don't get your hopes up 😂.

I'm thinking that for this PR, we focus on getting everything working without `rr`. And then in a future PR, we can add `rr` support,

File: .buildkite/platforms/linux64.yml [hunk 1]
```diff
@@ -70,7 +70,10 @@ steps:
       julia-artifact/bin/julia -e '@info "" Sys.CPU_THREADS'
 
       echo "--- Run the Julia test suite"
-      julia-artifact/bin/julia .buildkite/rr_capture.jl julia-artifact/bin/julia -e 'Base.runtests(["all"]; ncores = Sys.CPU_THREADS)'
+      
+      
+      
+      julia-artifact/bin/julia .buildkite/rr_capture.jl julia-artifact/bin/julia -e 'const HOME = joinpath(pwd(), "myhomedir"); mkpath(HOME); ENV["HOME"] = HOME; ENV["USER"] = "myusername"; Base.runtests(["all"]; ncores = Sys.CPU_THREADS)'
```
Comment (staticfloat) [inline .buildkite/platforms/linux64.yml:76]: I don't think you want to do it this way; instead you should just set `HOME: /home` in the YAML environment variable section.  That is an empty directory that should be fine to modify.

Comment (DilumAluthge) [reply .buildkite/platforms/linux64.yml]: Done!

**Comment by DilumAluthge** [comment]
I set `$HOME` to `/home`, but it looks like at some point it's being overridden to `/root`?

**Comment by staticfloat** [comment]
Okay so the test we're failing is one that explicitly nulls out `$HOME`.  Sigh.  I guess we'll add a user to the system.

**Comment by DilumAluthge** [comment]
😢 

Where do we need to add the user?

**Comment by DilumAluthge** [comment]
From the `tester_linux64` builder:
> ERROR: Unable to load dependent library /cache/build/amdci7-1/julialang/julia-master/julia-artifact/bin/../lib/julia/libjulia-internal.so.1
> 
> Message:/lib/x86_64-linux-gnu/libm.so.6: version `GLIBC_2.29' not found (required by /cache/build/amdci7-1/julialang/julia-master/julia-artifact/bin/../lib/julia/libjulia-internal.so.1)

**Comment by DilumAluthge** [comment]
But at least the `doctest` builder is passing now!

**Comment by DilumAluthge** [comment]
If we're going to have a user with a home directory in the `tester_linux64` builder, perhaps we don't need to set the `JULIA_DEPOT_PATH` environment variable?

**Comment by DilumAluthge** [comment]
```
Test Failed at /cache/build/amdci5-0/julialang/julia-master/julia-artifact/share/julia/test/file.jl:506
  Expression: chown(file, -2, -1)
    Expected: Base.IOError("chown(\"/tmp/jl_M6hcxb/bfile.txt\", -2, -1): operation not permitted (EPERM)", -1)
      Thrown: Base.IOError("chown(\"/tmp/jl_M6hcxb/bfile.txt\", -2, -1): invalid argument (EINVAL)", -22)
Test Failed at /cache/build/amdci5-0/julialang/julia-master/julia-artifact/share/julia/test/file.jl:507
  Expression: chown(file, -1, -2)
    Expected: Base.IOError("chown(\"/tmp/jl_M6hcxb/bfile.txt\", -1, -2): operation not permitted (EPERM)", -1)
      Thrown: Base.IOError("chown(\"/tmp/jl_M6hcxb/bfile.txt\", -1, -2): invalid argument (EINVAL)", -22)
```

I'm not sure what's going on here. Locally, I always get `EPERM`.

**Comment by DilumAluthge** [comment]
What's the correct way for me to set the `JULIA_SHELL` environment variables to `/bin/bash`?

**Comment by staticfloat** [comment]
@DilumAluthge Just so you know, in `.yml` files, you need to double-escape all your `$VAR` usages to look like `$$VAR`, otherwise what happens is that the `.yml` file gets interpolated at `.yml` submission time, rather than at script run time; this allows you to do things like `JOB_NAME=package_linux64 buildkite-agent pipeline upload package.yml` and all of the `$JOB_NAME` references within your `pipeline.yml` will be replaced with `package_linux64` before it even gets uploaded to buildkite.  So the values for `USER` and `HOME` that we're seeing in the logs when you reference them with a single `$` are actually the values for the agent that is _uploading_ `package_linux64.yml`.

**Comment by DilumAluthge** [comment]
Ahhhh, that explains a lot.

**Comment by DilumAluthge** [comment]
Okay, looks like we are making progress. Originally, we had 12 failures and 6 errors, and now we are down to 10 failures and 0 errors.

**Comment by staticfloat** [comment]
@Keno We're running Julia tests inside of `sandbox` here, and `chown()` is giving `EINVAL` instead of `EPERM` when we pass in invalid UIDs.  This is presumably because `/proc/self/uid_map` is `1000 1000 1`, meaning there is only one valid UID in the system.  What do you propose we do?  Should we try to mess with `/etc/subuid`, or should we detect this and disable the tests within the sandbox?

**Comment by Keno** [comment]
Seems inconsistent, but that's Linux for you. I'd just allow EINVAL too in that test.

**Comment by DilumAluthge** [comment]
I was thinking that it would be nice to keep this PR as Buildkite-changes-only. What do y'all think of making a separate PR to modify the tests, and then rebasing this PR once the test-modifying-PR is merged?

**Comment by DilumAluthge** [comment]
> I was thinking that it would be nice to keep this PR as Buildkite-changes-only. What do y'all think of making a separate PR to modify the tests, and then rebasing this PR once the test-modifying-PR is merged?

https://github.com/JuliaLang/julia/pull/41682

**Comment by DilumAluthge** [comment]
Any idea why these `cmdlineargs` tests are failing?

**Comment by DilumAluthge** [comment]
These `cmdlineargs` tests are really confusing:
```
      From worker 7:	┌ Info:
      From worker 7:	└   s = "ERROR: could not load library \"/cache/build/amdci5-2/julialang/julia-master/julia-artifact/share/julia/test/nonexistent\"\n/cache/build/amdci5-2/julialang/julia-master/julia-artifact/share/julia/test/nonexistent.so: cannot open shared object file: No such file or directory\n"
      From worker 7:	┌ Info:
      From worker 7:	└   p = Process(`/cache/build/amdci5-2/julialang/julia-master/julia-artifact/bin/julia -Cnative -J/cache/build/amdci5-2/julialang/julia-master/julia-artifact/lib/julia/sys.so --depwarn=error --check-bounds=yes -g1 --startup-file=no --sysimage=/cache/build/amdci5-2/julialang/julia-master/julia-artifact/share/julia/test/nonexistent`, ProcessSignaled(11))
      From worker 7:	┌ Info:
      From worker 7:	└   p.exitcode = 0
```

It's kind of strange that the process `p` exited with exit code zero, even though the `stderr` output (the string `s`) shows that it printed a fatal error.

**Comment by DilumAluthge** [comment]
Here's an update: the `tester_linux64` builder is failing with 7 failures total (6 fails and 1 error). All other jobs are passing.

My plan is:
1. Disable the `tester_linux64` builder.
2. Merge this PR.
3. Make a new PR to enable the `tester_linux64` builder.

**Comment by DilumAluthge** [comment]
I took all the TODOs that we did not complete in this PR, and I moved them to #41509.

**Pull request merged**